src/libsync/propagateupload.cpp

```diff
@@ -190,22 +194,53 @@ bool PollJob::finished()
     return true;
 }
 
-
 void PropagateUploadFileQNAM::start()
 {
     if (_propagator->_abortRequested.fetchAndAddRelaxed(0)) {
         return;
     }
 
+    const QString filePath = _propagator->getFilePath(_item._file);
+
+    // remember the modtime before checksumming to be able to detect a file
+    // change during the checksum calculation
+    _item._modtime = FileSystem::getModTime(filePath);
+
+    _stopWatch.start();
+
+    // do whatever is needed to add a checksum to the http upload request.
+    // in any case, the validator will emit signal startUpload to let the flow
+    // continue in slotStartUpload here.
+    _validator = new TransmissionChecksumValidator(filePath);
+    connect(_validator, SIGNAL(validated()), this, SLOT(slotStartUpload()));
+    _validator->uploadValidation( &_item );
+}
+
+void PropagateUploadFileQNAM::slotStartUpload()
+{
+    _validator->deleteLater();
+
     const QString fullFilePath(_propagator->getFilePath(_item._file));
 
     if (!FileSystem::fileExists(fullFilePath)) {
         done(SyncFileItem::SoftError, tr("File Removed"));
         return;
     }
+    _stopWatch.addLapTime(QLatin1String("Checksum"));
+
+    // Update the mtime and size, it might have changed since discovery.
+    time_t prevModtime = _item._modtime; // the value was set in PropagateUploadFileQNAM::start()
+    // but a potential checksum calculation could have taken some time during which the file could
+    // have been changed again, so better check again here.
 
     // Update the mtime and size, it might have changed since discovery.
     _item._modtime = FileSystem::getModTime(fullFilePath);
+    if( prevModtime != _item._modtime ) {
+        _propagator->_anotherSyncNeeded = true;
+        done(SyncFileItem::SoftError, tr("Local file changed while calculating the checksum."));
```

Comment:
"calculating the checksum" is an implementation details.  For an error which is likely to happen i think we should not let the user that.  Just maybe "File changed during sync" would be a better error message.